pkg/ffi2abi/ffi.go

```diff
@@ -0,0 +1,421 @@
+		schema.Type = fftypes.JSONAnyPtr(fmt.Sprintf(`"%s"`, objectType))
+		schema.Properties = make(map[string]*Schema, len(typeComponent.TupleChildren()))
+		for i, tupleChild := range typeComponent.TupleChildren() {
+			childSchema := getSchemaForABIInput(ctx, tupleChild)
+			childSchema.Details.Index = new(int)
+			*childSchema.Details.Index = i
+			schema.Properties[tupleChild.KeyName()] = childSchema
+		}
+	}
+	return schema
+}
+
+func convertFFIParamsToABIParameters(ctx context.Context, params fftypes.FFIParams) (abi.ParameterArray, error) {
+	abiParamList := make(abi.ParameterArray, len(params))
+	for i, param := range params {
+
+		// We need to validate the schema against the Ethereum FFI metaschema here
+		// For example, if someone is creating an event listener, we need to make
+		// sure we have all the correct ABI fields. This does not validate function
+		// arguments themselves. The blockchain connector or node will do that.
+
+		c := fftypes.NewFFISchemaCompiler()
+		v := &ParamValidator{}
+		c.RegisterExtension(v.GetExtensionName(), v.GetMetaSchema(), v)
+		err := c.AddResource(param.Name, strings.NewReader(param.Schema.String()))
+		if err != nil {
+			return nil, i18n.WrapError(ctx, err, signermsgs.MsgInvalidFFIDetailsSchema, param.Name)
+		}
+		_, err = c.Compile(param.Name)
+		if err != nil {
+			return nil, i18n.WrapError(ctx, err, signermsgs.MsgInvalidFFIDetailsSchema, param.Name)
+		}
+
+		var s *Schema
+		// Errors here are unchecked because they cannot be hit if the above JSON Schema validation passed
+		_ = json.Unmarshal(param.Schema.Bytes(), &s)
+		abiParameter, _ := processField(ctx, param.Name, s)
+
+		tc, err := abiParameter.TypeComponentTreeCtx(ctx)
+		if err != nil {
+			return nil, i18n.WrapError(ctx, err, signermsgs.MsgInvalidFFIDetailsSchema, param.Name)
+		}
+		if !inputTypeValidForTypeComponent(ctx, s.Type, tc) {
+			return nil, i18n.NewError(ctx, signermsgs.MsgInvalidFFIDetailsSchema, param.Name)
+		}
+
+		abiParamList[i] = abiParameter
+	}
+	return abiParamList, nil
+}
+
+func inputTypeValidForTypeComponent(ctx context.Context, inputType *fftypes.JSONAny, tc abi.TypeComponent) bool {
+	var inputTypeString string
+	if err := inputType.Unmarshal(ctx, &inputTypeString); err != nil {
+		if o, ok := inputType.JSONObjectOk(); ok {
+			if _, ok := o.GetObjectArrayOk("oneOf"); ok {
+				inputTypeString = integerType
+			}
+		}
+	}
+	switch inputTypeString {
+	case booleanType:
+		if tc.ElementaryType() == abi.ElementaryTypeBool {
+			return true
+		}
+	case integerType:
+		switch tc.ElementaryType() {
+		case abi.ElementaryTypeUint, abi.ElementaryTypeInt:
+			return true
+		}
+	case stringType:
+		switch tc.ElementaryType() {
+		case abi.ElementaryTypeAddress, abi.ElementaryTypeBytes, abi.ElementaryTypeString:
+			return true
+		}
+	case arrayType:
+		if tc.ArrayChild() != nil {
+			return true
+		}
+	case objectType:
+		if tc.TupleChildren() != nil {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
@peterbroadhurst if you can just double check to make sure I didn't miss anything here, this is the replacement for all the ugliness from the `ParamValidator` that we were talking about earlier this morning.

Reply:
Where does `ElementaryTypeFixed` go @nguyer ?
... see comment below, on whether the best place to solve this overall categorization on input/output is a new function on `ElementaryType`.

Reply:
I covered the `fixed`/`ufixed` types in https://github.com/hyperledger/firefly-signer/pull/11, but per the comment there there's a one-liner needed to handle `number` in the `firefly-common` code. I can create a PR for that too